kubernetes/cache.go

```diff
@@ -0,0 +1,458 @@
+package kubernetes
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/kiali/kiali/log"
+	"k8s.io/api/apps/v1beta1"
+	"k8s.io/api/apps/v1beta2"
+	batch_v1 "k8s.io/api/batch/v1"
+	batch_v1beta1 "k8s.io/api/batch/v1beta1"
+	"k8s.io/api/core/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/informers"
+	kube "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+)
+
+type (
+	// Inspired/reused from istio code:
+	// https://github.com/istio/istio/blob/master/mixer/adapter/kubernetesenv/cache.go
+	cacheController interface {
+		// Control Cache
+		Start()
+		HasSynced() bool
+		WaitForSync() bool
+		Stop()
+
+		// Business methods
+		GetCronJobs(namespace string) ([]batch_v1beta1.CronJob, error)
+		GetDeployment(namespace string, name string) (*v1beta1.Deployment, error)
+		GetDeployments(namespace string) ([]v1beta1.Deployment, error)
+		GetEndpoints(namespace, name string) (*v1.Endpoints, error)
+		GetJobs(namespace string) ([]batch_v1.Job, error)
+		GetPods(namespace string) ([]v1.Pod, error)
+		GetReplicationControllers(namespace string) ([]v1.ReplicationController, error)
+		GetReplicaSets(namespace string) ([]v1beta2.ReplicaSet, error)
+		GetService(namespace string, name string) (*v1.Service, error)
+		GetServices(namespace string) ([]v1.Service, error)
+		GetStatefulSet(namespace string, name string) (*v1beta2.StatefulSet, error)
+		GetStatefulSets(namespace string) ([]v1beta2.StatefulSet, error)
+	}
+
+	controllerImpl struct {
+		clientset       kube.Interface
+		refreshDuration time.Duration
+		stopChan        chan struct{}
+		syncCount       int
+		maxSyncCount    int
+		isErrorState    bool
+		lastError       error
+		lastErrorLock   sync.Mutex
+		controllers     map[string]cache.SharedIndexInformer
+	}
+)
+
+var (
+	lastCacheErrorLock sync.Mutex
+	errorCallbacks     []func(error)
+)
+
+func init() {
+	setupErrorHandlers()
+	errorCallbacks = make([]func(error), 0)
+}
+
+func setupErrorHandlers() {
+	nErrFunc := len(utilruntime.ErrorHandlers)
+	customErrorHandler := make([]func(error), nErrFunc+1)
+	for i, errorFunc := range utilruntime.ErrorHandlers {
+		customErrorHandler[i] = errorFunc
+	}
+	customErrorHandler[nErrFunc] = func(err error) {
+		for _, callback := range errorCallbacks {
+			callback(err)
+		}
+	}
+	utilruntime.ErrorHandlers = customErrorHandler
+}
+
+func registerErrorCallback(callback func(error)) {
+	defer lastCacheErrorLock.Unlock()
+	lastCacheErrorLock.Lock()
+	errorCallbacks = append(errorCallbacks, callback)
+}
+
+func newCacheController(clientset kube.Interface, refreshDuration time.Duration) cacheController {
+	newControllerImpl := controllerImpl{
+		clientset:       clientset,
+		refreshDuration: refreshDuration,
+		stopChan:        nil,
+		controllers:     initControllers(clientset, refreshDuration),
+		syncCount:       0,
+		maxSyncCount:    20, // Move this to config ? or this constant is good enough ?
+	}
+	registerErrorCallback(newControllerImpl.ErrorCallback)
```

Comment:
I think we should be able to unregister error callbacks when cache controller is stopped, no?
So, as they are stored in a global slice, maybe "register" could return the index in slice, and on "stop" it would remove from slice using that index?

Reply:
Actually when I write "when cache controller is stopped" I'm not sure about what "stop" means here. What I want to say is, when the controller reaches its end of life.

Reply:
+1
But note that this registration happens once per IstioClient, which has not a destroy/clean method an relies on the lifecycle of the app.
This can be better improved, but I guess that probably it can go in a following PR as we'll need to add the caches for istio objects, too.

Reply:
=> https://issues.jboss.org/browse/KIALI-1841